backend/app/Http/Controllers/CompanyController.php

```diff
@@ -4,15 +4,127 @@
+use Internships\Http\Requests\Api\FilterCompanies;
+use Internships\Http\Requests\Api\GetCompaniesRequest;
+use Internships\Http\Requests\Api\GetManagedCompaniesRequest;
+use Internships\Http\Requests\CompanyRequest;
+use Internships\Http\Resources\CityResource;
+use Internships\Http\Resources\CompanyResource;
+use Internships\Http\Resources\CompanySummaryResource;
+use Internships\Http\Resources\DepartmentResource;
+use Internships\Models\Company;
+use Internships\Models\Department;
+use Spatie\DataTransferObject\Exceptions\UnknownProperties;
 
 class CompanyController extends Controller
 {
-    public function create(Request $request): Response
+    public function index(GetCompaniesRequest $request): Response|array
+    {
+        return $this->list($request);
+    }
+
+    public function manage(): Response
+    {
+        return $this->list(new GetManagedCompaniesRequest());
+    }
+
+    public function create(): Response
     {
         return inertia(
             "Company/Create",
         );
     }
+
+    /**
+     * @throws UnknownProperties
+     */
+    public function store(CompanyRequest $request): RedirectResponse
+    {
+        $request->data();
+
+        return redirect()->route("company-manage")
+            ->with("success", "status.company_created");
+    }
+
+    /**
+     * @throws AuthorizationException
+     */
+    public function show(Company $company, GetCompaniesRequest $request): Response
+    {
+        $this->authorize("show", $company);
+        session(["view-source" => url()->previous()]);
+
+        return $this->list($request)->with(
+            "selectedCompany",
+            new CompanyResource($company),
+        );
+    }
+
+    public function close(): RedirectResponse
+    {
+        $source = session("view-source");
+
+        if (Str::is(route("index"), $source)
+            || Str::is(route("company-manage"), $source)
+            || Str::is(route("index") . "/?*", $source)
+            || Str::is(route("company-manage") . "/?*", $source)) {
+            return Redirect::to($source)->withInput();
+        }
+
+        return Redirect::to(route("index"));
+    }
+
+    /**
+     * @throws AuthorizationException
+     */
+    public function verify(Company $company): RedirectResponse
+    {
+        $this->authorize(Permission::ManageCompanies);
+
+        $company->update([
+            "status" => CompanyStatus::Verified,
+        ]);
+
+        return redirect()->route("company-manage")
+            ->with("success", "status.company_verified");
+    }
+
+    /**
+     * @throws AuthorizationException
+     */
+    public function delete(Company $company): RedirectResponse
+    {
+        $this->authorize("destroy", $company);
+        $company->delete();
+
+        return redirect()->route("company-manage")
+            ->with("success", "status.company_deleted");
+    }
+
+    protected function list(
+        GetCompaniesRequest|GetManagedCompaniesRequest $request,
```

Comment:
Why we have two request classes here?

Reply:
It's because the `list` function has shared functionality for two routes that list companies [index and manage]. It's more of a helper to avoid replicated code.

Both requests return Company Models, but their queries and authorization method are different.

Maybe I should move `list` function into request and return response instead of companies? I could make GetManagedCompaniesRequest extend GetCompaniesRequest.